torch/autograd/functional.py

```diff
@@ -0,0 +1,712 @@
+        ...   return x.exp().sum(dim=1)
+        >>> inputs = torch.rand(4, 4)
+        >>> v = torch.ones(4, 4)
+        >>> jvp(exp_reducer, inputs, v)
+        (tensor([6.3090, 4.6742, 7.9114, 8.2106]),
+         tensor([6.3090, 4.6742, 7.9114, 8.2106]))
+
+        >>> jvp(exp_reducer, inputs, v, create_graph=True)
+        (tensor([6.3090, 4.6742, 7.9114, 8.2106], grad_fn=<SumBackward1>),
+         tensor([6.3090, 4.6742, 7.9114, 8.2106], grad_fn=<SqueezeBackward1>))
+
+        >>> def adder(x, y):
+        ...   return 2 * x + 3 * y
+        >>> inputs = (torch.rand(2), torch.rand(2))
+        >>> v = (torch.ones(2), torch.ones(2))
+        >>> jvp(adder, inputs, v)
+        (tensor([2.2399, 2.5005]),
+         tensor([5., 5.]))
+    """
+
+    tuple_inputs, inputs = _as_tuple(inputs, "inputs", "jvp")
+    inputs = _grad_preprocess(inputs, create_graph=create_graph, need_graph=True)
+
+    if v is not None:
+        tuple_v, v = _as_tuple(v, "v", "jvp")
+        v = _grad_preprocess(v, create_graph=create_graph, need_graph=False)
+        _validate_v(v, inputs)
+    else:
+        if len(inputs) != 1 or inputs[0].nelement() != 1:
+            raise RuntimeError("The vector v can only be None if the input to the user-provided function "
+                               "is a single Tensor with a single element.")
+
+    outputs = func(*inputs)
+    tuple_outputs, outputs = _as_tuple(outputs, "outputs of the user-provided function", "jvp")
+    _check_requires_grad(outputs, "outputs", strict=strict)
+    grad_outputs = tuple(torch.zeros_like(out, requires_grad=True) for out in outputs)
+
+    grad_inputs = _autograd_grad(outputs, inputs, grad_outputs, create_graph=True)
+    _check_requires_grad(grad_inputs, "grad_inputs", strict=strict)
+
+    grad_res = _autograd_grad(grad_inputs, grad_outputs, v, create_graph=create_graph)
+
+    jvp = tuple()
+    for i, jvp_i in enumerate(grad_res):
+        if jvp_i is None:
+            if strict:
+                raise RuntimeError("The gradient with respect to the input is independent of entry {}"
+                                   " in the grad_outputs when using the double backward trick to compute"
+                                   " forward mode gradients. This is not allowed in strict mode.".format(i))
+            jvp_i = torch.zeros_like(outputs[i])
+        else:
+            if strict and create_graph and not jvp_i.requires_grad:
+                raise RuntimeError("The jacobian of the user-provided function is independent of "
+                                   "input {}. This is not allowed in strict mode when create_graph=True.".format(i))
+        jvp += (jvp_i,)
+
+    # Cleanup objects and return them to the user
+    outputs = _grad_postprocess(outputs, create_graph)
+    jvp = _grad_postprocess(jvp, create_graph)
+
+    return _tuple_postprocess(outputs, tuple_outputs), _tuple_postprocess(jvp, tuple_outputs)
+
+
+def jacobian(func, inputs, create_graph=False, strict=False):
```

Comment:
Hmm, I noticed that jax has jacfwd and jacrev. Do you have a sense of, if we do need and implement efficient forward mode AD, if we should have a split in the API like that?

Reply:
jax also has a `jax.jacobian` function. I think given that we don't have a forward mode, the general function is enough. If we add forward mode later, we can do like them and have all 3 functions (I think jax.jacobian is just the same as jax.jacrev)